conjure-codegen/src/objects/builder/mod.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright 2021 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use crate::context::{CollectionSetterBounds, CollectionType, Context, SetterBounds};
+use crate::types::{FieldDefinition, ObjectDefinition};
+use proc_macro2::{Ident, TokenStream};
+use quote::quote;
+use std::collections::HashSet;
+
+mod fallible;
+mod staged;
+
+pub fn generate(ctx: &Context, def: &ObjectDefinition) -> TokenStream {
+    if ctx.staged_builders() {
+        staged::generate(ctx, def)
+    } else {
+        fallible::generate(ctx, def)
+    }
+}
+
+fn field_names(ctx: &Context, def: &ObjectDefinition) -> HashSet<String> {
+    def.fields()
+        .iter()
+        .map(|f| ctx.field_name(f.field_name()).to_string())
+        .collect()
+}
+
+struct Setter {
+    name: Ident,
+    params: TokenStream,
+    args: Vec<SetterArg>,
+    where_: TokenStream,
+    op: SetterOp,
+}
+
+struct SetterArg {
+    name: TokenStream,
+    type_: TokenStream,
+}
+
+enum SetterOp {
+    Assign { rhs: TokenStream },
+    Call { call: TokenStream },
+}
+
+fn field_setters(
+    ctx: &Context,
+    def: &ObjectDefinition,
+    field: &FieldDefinition,
+    field_names: &HashSet<String>,
+) -> Vec<Setter> {
+    let name = ctx.field_name(field.field_name());
+
+    match ctx.setter_bounds(def.type_name(), field.type_(), quote!(#name)) {
+        SetterBounds::Simple {
+            argument_type,
+            assign_rhs,
+        } => {
+            vec![Setter {
+                params: quote!(),
+                args: vec![SetterArg {
+                    name: quote!(#name),
+                    type_: argument_type,
+                }],
+                where_: quote!(),
+                op: SetterOp::Assign { rhs: assign_rhs },
+                name,
+            }]
+        }
+        SetterBounds::Generic {
+            argument_bound,
+            assign_rhs,
+        } => {
+            vec![Setter {
+                params: quote!(<T>),
+                args: vec![SetterArg {
+                    name: quote!(#name),
+                    type_: quote!(T),
+                }],
+                where_: quote!(where T: #argument_bound),
+                op: SetterOp::Assign { rhs: assign_rhs },
+                name,
+            }]
+        }
+        SetterBounds::Collection {
```

Comment:
Factor out the business logic for this and other longer match cases.

Reply:
Refactored things a bit to clean up.